pkg/collector/corechecks/network-devices/cisco-sdwan/client/request.go

```diff
@@ -0,0 +1,172 @@
+
+	var bytes []byte
+	var statusCode int
+
+	for attempts := 0; attempts < client.maxAttempts; attempts++ {
+		err = client.authenticate()
+		if err != nil {
+			return nil, err
+		}
+
+		bytes, statusCode, err = client.do(req)
+
+		if err == nil && isValidStatusCode(statusCode) {
+			// Got a valid response, stop retrying
+			return bytes, nil
+		}
+	}
+
+	return nil, fmt.Errorf("http responsed with %d code", statusCode)
+}
+
+// get wraps client.get with generic type content and unmarshalling (methods can't use generics)
+func get[T Content](client *Client, endpoint string, params map[string]string) (*Response[T], error) {
+	bytes, err := client.get(endpoint, params)
+	if err != nil {
+		return nil, err
+	}
+
+	var data Response[T]
+
+	err = json.Unmarshal(bytes, &data)
+	if err != nil {
+		return nil, err
+	}
+
+	return &data, nil
+}
+
+func isValidStatusCode(code int) bool {
+	return code >= 200 && code < 400
+}
+
+// getMoreEntries gets all results from paginated endpoints
+func getMoreEntries[T Content](client *Client, endpoint string, params map[string]string, pageInfo PageInfo) ([]T, error) {
+	var responses []T
+	currentPageInfo := pageInfo
+
+	// Loop while API response indicates there is more entries
+	for page := 0; currentPageInfo.MoreEntries || currentPageInfo.HasMoreData; page++ {
+		// Error if max number of pages is reached
+		if page >= client.maxPages {
+			return nil, fmt.Errorf("max number of page reached, increase API count or max number of pages")
+		}
+
+		log.Tracef("Getting page %d from endpoint %s", page+1+1, endpoint)
+		// Update the params to get next API page
+		err := updatePaginationParams(currentPageInfo, params)
+		if err != nil {
+			return nil, err
+		}
+		log.Tracef("Pagination params for page %d from endpoint %s : %v", page+1+1, endpoint, params)
+
+		// Call the endpoint with the new params
+		data, err := get[T](client, endpoint, params)
+		if err != nil {
+			return nil, err
+		}
+
+		responses = append(responses, data.Data...)
+		currentPageInfo = data.PageInfo
+	}
+
+	return responses, nil
+}
+
+// updatePaginationParams updates query params to get next page
+func updatePaginationParams(info PageInfo, params map[string]string) error {
+	if info.MoreEntries {
+		// For endpoints that uses index-based pagination
+		endID, err := strconv.Atoi(info.EndID)
+		if err != nil {
+			return err
+		}
+		params["startId"] = fmt.Sprintf("%v", endID+1)
+	} else if info.HasMoreData {
+		// For endpoints that uses scroll-based pagination (ES like)
+		params["scrollId"] = info.ScrollID
+	}
+	return nil
+}
+
+// getAllEntries gets all entries from paginated endpoints
+func getAllEntries[T Content](client *Client, endpoint string, params map[string]string) (*Response[T], error) {
+	data, err := get[T](client, endpoint, params)
+	if err != nil {
+		return nil, err
+	}
+
+	// If API response is paginated, get the rest
+	entries, err := getMoreEntries[T](client, endpoint, params, data.PageInfo)
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** I assume some endpoints are paginated and some are not. Should we avoid calling `getMoreEntries` if there is no pagination?‏

Reply:
AFAIK all endpoints are paginated
`getMoreEntries` should be a noop if `HasMoreData` and `MoreEntries` page infos are false, do you think it would make more sense to check that before calling `getMoreEntries` ? No strong opinion on my side